### PR TITLE
Fix exception with aws credentials

### LINKED ISSUE #311: Bug on 2.8.0: Exception on reading the Jenkins settings for Bitbucket Cloud
Hi,

I am duplicating this issue from:
https://issues.jenkins-ci.org/browse/JENKINS-62180

I am getting an exception when I open Manage Jenkins -> Configuration after installing the newest version of Bitbucket Cloud 2.8.0.

```
Error while serving
https://system-dev-jenkins.build.mttnow.com/descriptorByName/com.cloudbees.
jenkins.plugins.bitbucket.endpoints.BitbucketCloudEndpoint/
fillCredentialsIdItems com.amazonaws.services.ecr.model.AmazonECRException: The
security token included in the request is invalid (Service: AmazonECR; Status
Code: 400; Error Code: UnrecognizedClientException; Request ID:
bb63ccfc-1924-475f-bf3e-9d219f70d6ac) at
com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(
AmazonHttpClient.java:1799) at
com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleServiceErrorResponse(
AmazonHttpClient.java:1383) at
com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(
AmazonHttpClient.java:1359) at
com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(
AmazonHttpClient.java:1139) at
com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.
java:796) at
com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(
AmazonHttpClient.java:764) at
com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.
java:738) at
com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.
java:698) at
com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(
AmazonHttpClient.java:680) at
com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:544) at
com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:524) at
com.amazonaws.services.ecr.AmazonECRClient.doInvoke(AmazonECRClient.java:2383)
at com.amazonaws.services.ecr.AmazonECRClient.invoke(AmazonECRClient.java:2350)
at com.amazonaws.services.ecr.AmazonECRClient.invoke(AmazonECRClient.java:2339)
at
com.amazonaws.services.ecr.AmazonECRClient.executeGetAuthorizationToken(
AmazonECRClient.java:1166) at
com.amazonaws.services.ecr.AmazonECRClient.getAuthorizationToken(AmazonECRClient
.java:1136) at
com.cloudbees.jenkins.plugins.amazonecr.AmazonECSRegistryCredential.getPassword(
AmazonECSRegistryCredential.java:131) at
com.cloudbees.jenkins.plugins.bitbucket.api.credentials.
BitbucketOAuthCredentialMatcher.matches(BitbucketOAuthCredentialMatcher.java:25)
at
com.cloudbees.plugins.credentials.matchers.AnyOfMatcher.matches(AnyOfMatcher.
java:67) at
com.cloudbees.plugins.credentials.CredentialsProvider.getCredentialIds(
CredentialsProvider.java:1196) at
com.cloudbees.plugins.credentials.CredentialsProvider.listCredentials(
CredentialsProvider.java:484) at
com.cloudbees.plugins.credentials.common.AbstractIdCredentialsListBoxModel.
includeMatchingAs(AbstractIdCredentialsListBoxModel.java:499) at
com.cloudbees.jenkins.plugins.bitbucket.endpoints.
AbstractBitbucketEndpointDescriptor.doFillCredentialsIdItems(
AbstractBitbucketEndpointDescriptor.java:61) at
java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:627) at
org.kohsuke.stapler.Function$MethodFunction.invoke(Function.java:396) Caused:
java.lang.reflect.InvocationTargetException at
org.kohsuke.stapler.Function$MethodFunction.invoke(Function.java:400)
```

This is the screen shot:
Screen Shot 2020-05-06 at 14 24 01

I believe it has something to do with a Amazon credentials provider and a few of the credentials I might have of that type, but not sure why this is happening.

It is related to the new OAuth feature - as the exception happens on BitbucketOAuthCredentialMatcher.

Thanks.

### PR DESCRIPTION
**Pull request by kurddt**
Should fix #311. It seems that aws credential throws an exception if the password can't be retrieved

### Your checklist for this pull request

- [X] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side) and not your master branch!
- [X] Ensure that the pull request title represents the desired changelog entry
- [X] Please describe what you did
- [X] Link to relevant issues in GitHub or in [Jenkins JIRA](https://issues.jenkins-ci.org)
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [ ] Did you provide a test-case? That demonstrates feature works or fixes the issue.

### CODE REVIEW

File: src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketOAuthCredentialMatcher.java [hunk 1]
```diff
@@ -19,13 +22,18 @@ public boolean matches(Credentials item) {
...
-        return !isEMail && validKeyLenght && validSecretLenght;
+        try {
+            UsernamePasswordCredentials usernamePasswordCredential = ((UsernamePasswordCredentials) item);
+            String username = usernamePasswordCredential.getUsername();
+            boolean isEMail = username.contains(".") && username.contains("@");
+            boolean validSecretLenght = Secret.toString(usernamePasswordCredential.getPassword()).length() == secretLenght;
+            boolean validKeyLenght = username.length() == keyLenght;
+
+            return !isEMail && validKeyLenght && validSecretLenght;
+        } catch(Exception e) {
```
Comment (jetersen) [inline src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketOAuthCredentialMatcher.java:33]: Seems a little broad.

Doesn't really seem like a fix. More like your just logging the exception.

Comment (kurddt) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketOAuthCredentialMatcher.java:33]: The idea here is to be able to continue, if some of the defined credentials can't get the password

Comment (jetersen) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketOAuthCredentialMatcher.java:33]: Than why log the exception at all?

I feel like we would be better suited creating a new credential class.

Comment (kurddt) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketOAuthCredentialMatcher.java:33]: That way, if the throwing credential should contains the OAuth key/secret then the user have a way to know what happened.

But I do agree for the new credential class, but I really don't know how to do that.

Comment (jetersen) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketOAuthCredentialMatcher.java:33]: Here is a few examples: https://github.com/search?q=org%3Ajenkinsci+extends+BaseStandardCredentials&type=Code

Comment (jetersen) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketOAuthCredentialMatcher.java:33]: Here are the docs: https://github.com/jenkinsci/credentials-plugin/blob/master/docs/implementation.adoc#implementing-a-new-credentials-type

Comment (bitwiseman) [reply src/main/java/com/cloudbees/jenkins/plugins/bitbucket/api/credentials/BitbucketOAuthCredentialMatcher.java:33]: @kurddt A new credential type was also created recently in the github-branch-source-plugin. You can look there for a concrete example.

**Comment by kurddt** [comment]
@jetersen I'm working on using a dedicated credential class, I think it's best to close (reject) this PR and I will make a new one. What do you think ?

**Comment by kurddt** [comment]
@bitwiseman If credentials do not match, they will not appear on the dropdown menu when creating a source. If no other "matching" credential is found then only the default `-None-` credential will be available. I'm not sure what the behaviour will be then

**Pull request closed without merging**